### PR TITLE
feat(utils): add ReduceAsync function

### PR DESCRIPTION
**Pull request by nathan-pichon**
### What does it do?

It adds a reduce function for async utils.

### Why is it needed?

To be able to do a reduce over a map of value in an async environment (having promises inside the reduce function).

### How to test it?

You can use it in the backend code or simply do a unit test as the function is not yet used in the codebase.

### Related issue(s)/PR(s)

https://github.com/strapi/strapi/pull/15134

### CODE REVIEW

File: packages/core/utils/lib/__tests__/async.test.js [hunk 1]
```diff
@@ -77,4 +77,48 @@ describe('Async utils', () => {
       expect(maxOperations).toEqual(2);
     });
   });
+  describe('reduceAsync', () => {
+    test('Should return a incremented number', async () => {
```
Comment (petersg83) [inline packages/core/utils/lib/__tests__/async.test.js:81]:
```suggestion
    test('Should return an incremented number', async () => {
```

File: packages/core/utils/lib/__tests__/async.test.js [hunk 2]
```diff
@@ -77,4 +77,48 @@ describe('Async utils', () => {
...
+
+      const reduceFunc = reduceAsync(numberPromiseArray);
+
+      await expect(async () => {
+        await reduceFunc(() => {
+          throw new Error('test');
+        }, null);
+      }).rejects.toThrow('test');
+    });
+    test('Should throw an error 2', async () => {
```
Comment (petersg83) [inline packages/core/utils/lib/__tests__/async.test.js:114]: Do you see a test name that would be more explicit on what is tested? Same for above

**Review by petersg83 — CHANGES_REQUESTED** [review]
noice

File: packages/core/utils/lib/async.js [hunk 3]
```diff
@@ -20,6 +20,17 @@ function pipeAsync(...methods) {
...
+/**
+ * @type { import('./async').ReduceAsync }
+ */
+const reduceAsync = curry(async (mixedArray, iteratee, initialValue) => {
+  let acc = initialValue;
+  for (let i = 0; i < mixedArray.length; i += 1) {
+    acc = await iteratee(acc, await mixedArray[i], i);
+  }
+  return acc;
+}, 2);
```
[the remark this answers is not in the document]
Comment (nathan-pichon) [reply packages/core/utils/lib/async.js:32]: There is a second argument even with `lodash/fp` https://lodash.com/docs/4.17.15#curry (`LodashCurryN` in TS definitions)
It's mandatory as the `initialValue` is optional.

File: packages/core/utils/lib/__tests__/async.test.js (hunk 2 above)
Comment (nathan-pichon) [reply packages/core/utils/lib/__tests__/async.test.js]: Should be good now 🙂

File: packages/core/utils/lib/async.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/utils/lib/async.js:32]: I read this in the [doc](https://github.com/lodash/lodash/wiki/FP-Guide):
Capture d’écran 2023-02-14 à 09 58 34

Is it curryN that would be used instead?
I just tried without the initial value and it doesn't seem to execute the function.
A test could be added to test that usecase?

File: packages/core/utils/lib/__tests__/async.test.js (hunk 2 above)
Comment (petersg83) [reply packages/core/utils/lib/__tests__/async.test.js]: Nice, I understand better thanks :)

**Review by petersg83 — CHANGES_REQUESTED** [review]

**Comment by nathan-pichon** [comment]
Didn't know why I cannot respond to your comments @petersg83, but thanks for the catch on curry/curryN. Added a test and use curryN now ! 👍

**Review by petersg83 — APPROVED** [review]
LGTM :)

**Review by Marc-Roig — APPROVED** [review]

**Pull request merged**